feat(transformer/class-properties): transform super member expressions that are inside static prop initializer

Dunqing opened the pull request:
This PR support for transforming `super.prop` to  `babelHelpers.superPropGet(_B, "prop", _B)`

Input:
```js
class A {
  static prop = 1;
}

class B extends A {
  static prop = 2;
  static propA = super.prop;
  static getPropA = () => super.prop;
}
```

Output:
```js
var _B;
class A {}
babelHelpers.defineProperty(A, "prop", 1);
class B extends A {}
_B = B;
babelHelpers.defineProperty(B, "prop", 2);
babelHelpers.defineProperty(B, "propA", babelHelpers.superPropGet(_B, "prop", _B));
babelHelpers.defineProperty(B, "getPropA", () => babelHelpers.superPropGet(_B, "prop", _B));
```

overlookmotel commented on crates/oxc_transformer/src/es2022/class_properties/supers.rs:
```diff
@@ -0,0 +1,76 @@
+    // or `Expression::ComputedMemberExpression`.
+    #[inline]
+    pub fn transform_member_expression(
+        &mut self,
+        expr: &mut Expression<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        let member = expr.to_member_expression();
+        if matches!(member.object(), Expression::Super(_)) {
+            self.transform_member_expression_impl(expr, ctx);
+        }
+    }
+
+    fn transform_member_expression_impl(
+        &mut self,
+        expr: &mut Expression<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        let property = match expr {
+            // `super.prop` -> `"prop"`
+            Expression::StaticMemberExpression(member) => {
+                let property = &member.property;
+                ctx.ast.expression_string_literal(
+                    property.span,
+                    property.name.clone(),
+                    Some(property.name.clone()),
+                )
+            }
+            // `super[expr]` -> `expr`
+            Expression::ComputedMemberExpression(member) => {
+                ctx.ast.move_expression(&mut member.expression)
+            }
+            Expression::PrivateFieldExpression(_) => {
+                unreachable!("`super` cannot access private fields")
+            }
+            _ => return,
+        };
+
+        *expr = self.create_super_prop_get(expr.span(), property, ctx);
+    }
```
It would be better to have 2 different visitors here for `StaticMemberExpression` and `ComputedMemberExpression`. Currently you funnel both through `transform_member_expression` and then you branch on what kind of `Expression` it is again in `transform_member_expression_impl`. This is unnecessary work.

Dunqing replied:
Thanks for pointing this out, we should do that!

overlookmotel commented on crates/oxc_transformer/src/es2022/class_properties/supers.rs:
```diff
@@ -0,0 +1,76 @@
+        property: Expression<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> Expression<'a> {
+        let class_binding = self.class_bindings.get_or_init_temp_binding(ctx);
```
This needs to use `ClassProperties::get_temp_binding` instead (defined in `static_prop_init.rs`, you'll need to make that method `pub(super)`).

```suggestion
        let class_binding = self.get_temp_binding(ctx);
```

It's really annoying we have to do this, and it's not at all obvious. The problem is that we store `class_bindings` in 2 places and which is the "source of truth" depends on whether the class has private properties or not. This is very bad design (my fault), but I couldn't find a better way at present.

overlookmotel commented on crates/oxc_transformer/src/es2022/class_properties/supers.rs:
```diff
@@ -0,0 +1,76 @@
+use super::ClassProperties;
+
+impl<'a, 'ctx> ClassProperties<'a, 'ctx> {
+    /// Transform member expression where object is a super.
```
Nit:

```suggestion
    /// Transform member expression where object `super`.
```

Dunqing commented:
Thank @overlookmotel for reviewing, now it more clearer!

overlookmotel reviewed (approved):
Pushed 4 nitty commits. @Dunqing If you're OK with them, please merge.

Dunqing committed:
feat(transformer/class-properties): transform super member expressions that are inside static prop initializer (#7815)

This PR support for transforming `super.prop` to  `babelHelpers.superPropGet(_B, "prop", _B)`

Input:
```js
class A {
  static prop = 1;
}

class B extends A {
  static prop = 2;
  static propA = super.prop;
  static getPropA = () => super.prop;
}
```

Output:
```js
var _B;
class A {}
babelHelpers.defineProperty(A, "prop", 1);
class B extends A {}
_B = B;
babelHelpers.defineProperty(B, "prop", 2);
babelHelpers.defineProperty(B, "propA", babelHelpers.superPropGet(_B, "prop", _B));
babelHelpers.defineProperty(B, "getPropA", () => babelHelpers.superPropGet(_B, "prop", _B));
```
